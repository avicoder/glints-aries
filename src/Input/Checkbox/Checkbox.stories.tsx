import * as React from 'react';
import { Story, Meta } from '@storybook/react';

import { Checkbox, CheckboxProps } from './Checkbox';
import { BaseContainer } from '../../Layout/GlintsContainer/GlintsContainer';

import styled from 'styled-components';

const TestBox = styled.div<{ fontSize: string }>`
  div {
    font-size: ${({ fontSize }) => fontSize};
  }

  &:not(:last-child) {
    margin-bottom: 4px;
  }
`;

export default {
  title: 'General/Checkbox',
  component: Checkbox,
  decorators: [Story => <BaseContainer>{Story()}</BaseContainer>],
} as Meta;

const Template: Story<CheckboxProps> = args => <Checkbox {...args} />;

export const Interactive = Template.bind({});
Interactive.args = {
  id: 'software-engineer',
  value: 'software-engineer',
  label: 'Software Engineer',
};

export const Bordered = Template.bind({});
Bordered.args = {
  id: 'bordered-engineer',
  value: 'bordered-engineer',
  label: 'Bordered Engineer',
  border: true,
};

export const Sizes = Template.bind({});
Sizes.args = {
  id: 'large-engineer',
  value: 'large-engineer',
  label: 'Large Engineer',
  size: 'large',
};

const FontSizeStory: Story<CheckboxProps> = args => (
  <div>
    <TestBox fontSize="10px">
      <Checkbox {...args} checked={true} id="10px" value="10px" label="10px" />
    </TestBox>
    <TestBox fontSize="14px">
      <Checkbox {...args} checked={true} id="14px" value="14px" label="14px" />
    </TestBox>
    <TestBox fontSize="18px">
      <Checkbox {...args} checked={true} id="18px" value="18px" label="18px" />
    </TestBox>
    <TestBox fontSize="48px">
      <Checkbox {...args} checked={true} id="48px" value="48px" label="48px" />
    </TestBox>
  </div>
);
export const FontSize = FontSizeStory.bind({});
FontSize.parameters = {
  docs: {
    description: {
      story: `If you change the checkbox root div's font-size, the entire checkbox will adjust its size accordingly.`,
    },
  },
};
