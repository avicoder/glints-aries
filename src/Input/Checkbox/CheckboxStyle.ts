import styled from 'styled-components';
import { SecondaryColor } from '../../Utils/Colors';
import { CheckboxProps } from './Checkbox';

export const CheckboxContainer = styled.div<CheckboxProps>`
  position: relative;
  display: inline-flex;
  font-size: ${({ size }) => (size === 'small' ? '14px' : '16px')};
  line-height: 1.5;

  &:focus {
    outline: none;
  }

  input {
    padding: 0;
    height: initial;
    width: initial;
    margin-bottom: 0;
    display: none;
    cursor: pointer;

    &:checked + label:after {
      content: '';
      display: block;
      position: absolute;
      top: 0.2em;
      left: 0.5em;
      width: 0.4em;
      height: 0.85em;
      border: solid ${SecondaryColor.white};
      border-width: 0 0.15em 0.15em 0;
      transform: rotate(45deg);
      ${({ border }) => {
        if (border) {
          return `
            margin-top: 10px;
            margin-left: 15px;
          `;
        }
      }};
    }

    &:checked + label:before {
      background: ${({ border }) =>
        border
          ? `${SecondaryColor.actionblue}`
          : `${SecondaryColor.darkgreen}`};
      border: 2px solid
        ${({ border }) =>
          border
            ? `${SecondaryColor.actionblue}`
            : `${SecondaryColor.darkgreen}`};
    }
  }

  label {
    position: relative;
    display: flex;
    justify-content: center;
    line-height: 1.5;
    cursor: pointer;
    outline: none;
    ${({ border }) => {
      if (border) {
        return `
          border: 1px solid #aaaaaa;
          height: ${({ size }: any) => (size === 'small' ? '40px' : '43px')};
          cursor: pointer;
          border-radius: 2px;
          padding: 10px 15px;
          &:hover {
            background: rgba(1, 126, 183, 0.1);
            border-color: ${SecondaryColor.actionblue};
          }
        `;
      }
    }};

    ${({ checked }) => {
      if (checked) {
        return `
          border-color: ${SecondaryColor.actionblue};
        `;
      }
    }};

    &:before {
      content: '';
      appearance: none;
      background-color: transparent;
      border: 2px solid ${SecondaryColor.lightblack};
      padding: 0.6em;
      display: inline-block;
      position: relative;
      vertical-align: middle;
      cursor: pointer;
      margin-right: 10px;
      border-radius: 2px;
      height: 18px;
      width: 18px;
    }
  }
`;
