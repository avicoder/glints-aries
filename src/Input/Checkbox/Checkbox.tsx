import * as React from 'react';
import classNames from 'classnames';

import { CheckboxContainer } from './CheckboxStyle';
import { isUndefined } from 'lodash';

const Checkbox: React.FunctionComponent<CheckboxProps> = ({
  id,
  label,
  value,
  onChange,
  size = 'small',
  border = false,
  className,
  checked,
  disabled = false,
  ...restProps
}: CheckboxProps) => {
  const [internalChecked, setInternalChecked] = React.useState(false);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInternalChecked(internalChecked => !internalChecked);
    if (onChange !== undefined) {
      return onChange(e);
    }
  };

  const combinedChecked = isUndefined(checked) ? internalChecked : checked;

  return (
    <CheckboxContainer
      className={classNames('aries-checkbox', className)}
      role="checkbox"
      aria-labelledby={id}
      aria-checked={combinedChecked}
      tabIndex={0}
      size={size}
      border={border}
      checked={combinedChecked}
      disabled={disabled}
    >
      <input
        type="checkbox"
        id={id}
        value={value}
        // onClick={handleClick}
        onChange={handleChange}
        checked={combinedChecked}
        disabled={disabled}
        {...restProps}
      />
      <label htmlFor={id} tabIndex={-1}>
        {label || value}
      </label>
    </CheckboxContainer>
  );
};

type HTMLInputProps = Omit<React.HTMLProps<HTMLInputElement>, 'size' | 'label'>;

export interface CheckboxProps extends HTMLInputProps {
  label?: React.ReactNode;
  size?: 'large' | 'small';
  border?: boolean;
  value?: string;
  disabled?: boolean;
}

export default Checkbox;
