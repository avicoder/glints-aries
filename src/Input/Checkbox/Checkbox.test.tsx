import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import { fireEvent, render } from '@testing-library/react';

import Checkbox, { CheckboxProps } from './Checkbox';

const props = {
  id: 'software-engineer',
  value: 'Software Engineer',
  onClick: jest.fn().mockImplementation(event => event.target.value),
};

function setupCheckbox(props: CheckboxProps) {
  const { id, value, onClick, ...restProps } = props;
  const { getByText, getByLabelText, getByRole, ...utils } = render(
    <Checkbox id={id} value={value} onClick={onClick} {...restProps} />
  );
  const checkboxInput = getByLabelText(value as string) as HTMLInputElement;
  const checkboxLabel = getByText(value as string) as HTMLLabelElement;
  const checkboxContainer = getByRole('checkbox');
  return { checkboxInput, checkboxLabel, checkboxContainer, utils };
}

afterEach(() => {
  props.onClick.mockClear();
});

it(`<Checkbox> should render an input with id, value and onClick props and a label with the text ${props.value}`, () => {
  const CheckboxSnapshot = renderer
    .create(
      <Checkbox id={props.id} value={props.value} onClick={props.onClick} />
    )
    .toJSON();
  expect(CheckboxSnapshot).toMatchSnapshot();
});

it('when toggling checkbox, it should fire onClick once and become checked, then fire onClick once and become unchecked', () => {
  const { checkboxInput, checkboxLabel } = setupCheckbox(props);

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(true);

  props.onClick.mockClear();

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(false);
});

it('when controlled, it should fire onClick but remain unchanged', () => {
  const { checkboxInput, checkboxLabel } = setupCheckbox({
    ...props,
    checked: false,
  });

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(false);
});

it('when in mixed syaye, it should fire onClick but remain unchanged', () => {
  const { checkboxInput, checkboxLabel } = setupCheckbox({
    ...props,
    checked: false,
    indeterminate: true,
  });

  fireEvent.click(checkboxLabel);
  expect(props.onClick).toHaveBeenCalledTimes(1);
  expect(checkboxInput.checked).toEqual(true);
});

it('when controlled, the component should change with checked prop', () => {
  const Component = (props: CheckboxProps) => <Checkbox {...props} />;

  const { rerender, getByLabelText } = render(
    <Component {...props} checked={false} />
  );

  const checkboxInput = getByLabelText(
    props.value as string
  ) as HTMLInputElement;
  expect(checkboxInput.checked).toEqual(false);

  rerender(<Component checked={true} />);

  expect(checkboxInput.checked).toEqual(true);
});

it('when passed label, it should render the label', () => {
  const label = <p data-testid="label-child">label child</p>;
  const { getByTestId } = render(
    <Checkbox id={props.id} value={props.value} label={label} />
  );
  const LabelChild = getByTestId('label-child');
  expect(document.querySelector('label')).toContainElement(LabelChild);
});

describe('when it is rendered', () => {
  it('should display the correct label', () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput).toBeTruthy();
  });

  it('should display an unchecked checkbox', () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput.checked).toEqual(false);
  });

  it('should display a checked checkbox if checked=true', () => {
    const { checkboxInput } = setupCheckbox({ ...props, checked: true });
    expect(checkboxInput.checked).toEqual(true);
  });

  it('should display an unchecked checkbox if checked=false', () => {
    const { checkboxInput } = setupCheckbox({ ...props, checked: false });
    expect(checkboxInput.checked).toEqual(false);
  });

  it('should report as checked if intermediate=true', () => {
    const { checkboxInput } = setupCheckbox({ ...props, indeterminate: true });
    expect(checkboxInput.checked).toEqual(true);
  });

  it('should have attribute aria-checked=false if checked=false', () => {
    const { checkboxContainer } = setupCheckbox({
      ...props,
      checked: false,
    });
    expect(checkboxContainer.getAttribute('aria-checked')).toEqual('false');
  });

  it('should have attribute aria-checked=true if checked=true', () => {
    const { checkboxContainer } = setupCheckbox({
      ...props,
      checked: true,
    });
    expect(checkboxContainer.getAttribute('aria-checked')).toEqual('true');
  });

  it('should have attribute aria-checked=mixed if intermediate=true', () => {
    const { checkboxContainer } = setupCheckbox({
      ...props,
      indeterminate: true,
    });
    expect(checkboxContainer.getAttribute('aria-checked')).toEqual('mixed');
  });

  it('should have the correct input id and value', async () => {
    const { checkboxInput } = setupCheckbox(props);
    expect(checkboxInput.id).toEqual(props.id);
    expect(checkboxInput.value).toEqual(props.value);
  });

  it('should return the input value when onClick is passed a function: event => event.target.value', () => {
    const { checkboxLabel } = setupCheckbox(props);
    fireEvent.click(checkboxLabel);
    expect(props.onClick.mock.results[0].value).toEqual(props.value);
  });

  it('should return the input value when onClick is passed a function: event => event.target.value if indeterminate=true', () => {
    const { checkboxLabel } = setupCheckbox({ ...props, indeterminate: true });
    fireEvent.click(checkboxLabel);
    expect(props.onClick.mock.results[0].value).toEqual(props.value);
  });
});
