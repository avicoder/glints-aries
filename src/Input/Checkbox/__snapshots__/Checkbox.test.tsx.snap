// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> should render an input with id, value and onClick props and a label with the text Software Engineer 1`] = `
<div
  aria-checked={false}
  aria-labelledby="software-engineer"
  checked={false}
  className="CheckboxStyle__CheckboxContainer-c3ckvs-0 kydnPF aries-checkbox"
  role="checkbox"
  size="small"
  tabIndex={0}
>
  <input
    id="software-engineer"
    onClick={[Function]}
    type="checkbox"
    value="Software Engineer"
  />
  <label
    htmlFor="software-engineer"
    tabIndex={-1}
  >
    Software Engineer
  </label>
</div>
`;
